doc/manual/src/contributing/testing.md

```diff
@@ -86,6 +86,33 @@ GNU gdb (GDB) 12.1
 One can debug the Nix invocation in all the usual ways.
 For example, enter `run` to start the Nix invocation.
 
+### Characterization testing
+
+Occasionally, Nix utilizes a technique called [Characterization Testing](https://en.wikipedia.org/wiki/Characterization_test) as part of the functional tests.
+This technique is to include the exact output/behavior of a former version of Nix in a test in order to check that Nix continues to produce the same behavior going forward.
+
+For example, this technique is used for the language tests (`test/lang.sh`).
+
+It is frequently useful to regenerate the expected output.
+To do that, rerun the failed test with `_NIX_TEST_ACCEPT=1`.
+(At least, this is the convention we've used for `tests/lang.sh`.
+If we add more characterization testing we should always strive to be consistent.)
+
+An interesting situation to document is the case when these tests are "overfitted".
+For example, the language tests store both the expected result in the case of success, and expected diagnostics (errors and warnings).
+The expected successful output is supposed to be highly stable --- we do not intend to make breaking changes to the (stable parts of) the Nix language.
+However, the errors and warnings are not stable in this way, we are free to change how they are displayed at any time.
+
+Why bother to track these diagnostics if they are not stable then?
+The answer is that just because we Nix developers are free to change them whenever we like doesn't mean we don't care about them.
+On the contrary, errors and warnings are incredibly important to users.
+By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are code review.
+Regressions are caught, and improvements always show up in code review.
```

Comment:
```suggestion
While diagnostic outputs are not a stable interface, they are important to users.
By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are under code review.
Regressions are caught, and improvements always show up in code review.
```

Reply:
I am worried that the new concise version rapidly goes over central point.

This rhetorical question was basically what @edolstra was asking, and I think it is a good question: anything that is not unstable is not normative behavior, so why the hell has we testing it? (emphasizing "not normative" is probably better than "not stable", I am realizing now.)

If you think the rhetorical question is too chatty, we can pick a different form, like

> It may be surprising that we would test non-normative behavior like diagnostic outputs.
> After all, the point of tests is usually to pin down just normative.
> If tests pinned down all behavior, normative and non-normative alike, they would just be a second copy of the implementation!

(but perhaps shorter!)

The point is not to write a question per se, but to get the readers attention --- whether the reader is already surprised or not. This is the key point, and we don't want them to miss it.

Reply:
On a meta point, while one might argue the main portion of the manual is strictly reference not explanation, I think the contributing chapter is free to cover all quadrants, as it currently does, as `nix.dev` etc. will not be covering contributing to Nix.

Reply:
> It may be surprising that we would test non-normative behavior like diagnostic outputs.

This one's good for prepending it to my suggestion. I'm in general against being chatty in documentation.

Reply:
```suggestion
It may be surprising that we would test non-normative behavior like diagnostic outputs.
While diagnostic outputs are not a stable interface, they are important to users.
By recording the expected output, the test suite guards against accidental changes, and ensure the *result* (not just the code that implements it) of the diagnostic code paths are under code review.
Regressions are caught, and improvements always show up in code review.
```
OK!